UpComingMovies/app/src/main/java/com/example/douglas/upcomingmovies/MainActivity.java

```diff
@@ -0,0 +1,221 @@
+
+import com.example.douglas.upcomingmovies.model.Movie;
+import com.example.douglas.upcomingmovies.utilities.JsonHandler;
+import com.example.douglas.upcomingmovies.utilities.NetworkUtils;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.List;
+
+public class MainActivity extends AppCompatActivity {
+
+    private RecyclerView mRecyclerView;
+    private TextView mErrorMessageDisplay;
+    private MovieServiceAdapter mMovieServiceAdapter;
+    private int lastPageRequested=1;
+    private final int MAX_PAGES=10;
+    private List<Movie> fetchedMovies;
+
+    /**
+     * This method will create everything needed for start the application.
+     * @param savedInstanceState
+     */
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        /**
+         * Recycler view make the action of calling views easier and it is more efficient than
+         * call findViewById everytime. It makes the application more smooth maybe even more
+         * user-friendly.
+         */
+        mRecyclerView = (RecyclerView) findViewById(R.id.rv_movieservice);
+        mErrorMessageDisplay = (TextView) findViewById(R.id.tv_error_message_display);
+
+        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
+        mRecyclerView.setLayoutManager(layoutManager);
+
+        mRecyclerView.setHasFixedSize(true);
+
+        mMovieServiceAdapter = new MovieServiceAdapter();
+
+        /**
+         * The number max of pages was locked for avoid overflow of data. This could be
+         * changed for a future release. Therefore this code means that everytime the user
+         * reach(by scrolling) the last movie, the application will do one more request.
+         */
+        mMovieServiceAdapter.setOnBottomReachedListener(new MovieServiceAdapter.OnBottomReachedListener() {
+            @Override
+            public void onBottomReached(int position) {
+                if(lastPageRequested < MAX_PAGES) {
+                    lastPageRequested+=1;
+                    loadMovieData();
+                }
+            }
+        });
+
+        /**
+         * The adapter is the component able to create new items and populates them on the screen
+         * with data.
+         */
+        mRecyclerView.setAdapter(mMovieServiceAdapter);
+
+        loadMovieData();
+    }
+
+    /**
+     * This AsyncTask is almost a must have because the application need to do all the computations
+     * before the draw of the next frame, creating a AsysncTask is a modular aproach leaving the
+     * application free to be rendered while the content is being fetched.
+     */
+    public class UpComingMovieServiceTask extends AsyncTask<URL,Void,List<Movie>>{

```

Comment:
Any particular reason to keep the AsyncTask within the activity?

Reply:
I am aware that this can cause memory efficiency problems(rotating the phone for ex.), but I thought that I could use a Loader for the next release of the app, that could easily solve this problem, with just a few changes. I would just need to create a Loader ID, fill in the loader callbacks and initialize the loader with a LoaderManager. However I recognize that leaving the AsyncTask as it is was a mistake, I should keep it out of the MainActivity, reducing the problem and for future releases, I would implement the Loaders approach.